### PR TITLE
#7 junit5 test

### PR DESCRIPTION
**Pull request by Aayushchy**
Junit5 Test Cases for UserService and AdminService class

### CODE REVIEW

File: src/test/java/com/esewa/usermanagement/service/AdminServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    void testRegisterMultipleUsers_throwsException_forBlacklistedUser() {
+
+        List<User> users = TestUserBuilder.createUsers();
+
+        doNothing().when(emailService).sendEmail(any(User.class));
+        doNothing().when(registrationLogService).saveLog(any(RegistrationLog.class));
+        when(userService.registerUserAsync(any(User.class)))
+                .thenAnswer(invocation -> {
+                    User user = invocation.getArgument(0);
+                    return !"shivam".equals(user.getName())
```
Comment (devintern-esewa) [inline src/test/java/com/esewa/usermanagement/service/AdminServiceTest.java:72]: For constants like "shivam", can you put this as a field of this class?

`public static final String BLACKLISTED_USER = "shivam";`

File: src/test/java/com/esewa/usermanagement/service/UserServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    @Test
+    void testRegisterUserAsync_blacklistedUser() {
+
+        User blackListedUser = TestUserBuilder.createBlackListedUser();
+
+        when(userRepository.findByName(blackListedUser.getName()))
+                .thenReturn(Optional.empty());
+        when(userRepository.save(blackListedUser)).then(returnsFirstArg());
+
+        assertThrows(RuntimeException.class, () -> userService.registerUserAsync(blackListedUser));
```
Comment (devintern-esewa) [inline src/test/java/com/esewa/usermanagement/service/UserServiceTest.java:82]:
```suggestion
        RuntimeException ex = assertThrows(RuntimeException.class, () -> userService.registerUserAsync(blackListedUser));
        assertEquals("Expected error message", ex.getMessage());
```

**Review by devintern-esewa — CHANGES_REQUESTED** [review]
Verified with feedback

File: src/main/java/com/esewa/usermanagement/service/EmailServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                    }, () -> {
+                        log.error("Error while sending email. Unable to find template for {}", EmailMessage.REGISTRATION.getTemplateMessage());
+                        throw new RuntimeException(String.format("Error while sending email. Unable to find template for %s", EmailMessage.REGISTRATION.getTemplateMessage()));
+                    });
+
+            //javaMailSender.send(mailMessage);
+            emailLogRepository.save(new EmailLog(userDetail.getEmail(), EmailStatus.SENT, mailMessage.getText()));
+        } catch (Exception exception) {
+            log.error(exception.getMessage());
+            emailLogRepository.save(new EmailLog(userDetail.getEmail(), EmailStatus.FAILED, mailMessage.getText()));
```
[the remark this answers is not in the document]
Comment (Aayushchy) [reply src/main/java/com/esewa/usermanagement/service/EmailServiceImpl.java:57]: Done

File: src/main/java/com/esewa/usermanagement/service/EmailServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import org.springframework.mail.SimpleMailMessage;
+import org.springframework.mail.javamail.JavaMailSender;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Service;
+
+import java.util.Map;
+
+@Slf4j
+@Service
+public class EmailServiceImpl implements EmailService {
```
[the remark this answers is not in the document]
Comment (Aayushchy) [reply src/main/java/com/esewa/usermanagement/service/EmailServiceImpl.java:20]: Done

File: src/main/java/com/esewa/usermanagement/service/EmailServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    @Async
+    @Override
+    public void sendEmail(User userDetail) {
+
+        SimpleMailMessage mailMessage = new SimpleMailMessage();
+        mailMessage.setFrom(senderEmail);
+        mailMessage.setTo(userDetail.getEmail());
+        try {
+            emailTemplateRepository.findById(EmailMessage.REGISTRATION.getTemplateCode())
```
[the remark this answers is not in the document]
Comment (Aayushchy) [reply src/main/java/com/esewa/usermanagement/service/EmailServiceImpl.java:42]: Added Jedis cache implementation to cache template message

File: src/main/java/com/esewa/usermanagement/service/EmailServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                        String message = compileMessage(emailTemplate.getMessage(), Map.of("user", userDetail.getName(), "username", userDetail.getName()));
+                        log.info(message);
+                        mailMessage.setSubject(emailTemplate.getSubject());
+                        mailMessage.setText(message.replace("{username}", userDetail.getName()));
+                    }, () -> {
+                        log.error("Error while sending email. Unable to find template for {}", EmailMessage.REGISTRATION.getTemplateMessage());
+                        throw new RuntimeException(String.format("Error while sending email. Unable to find template for %s", EmailMessage.REGISTRATION.getTemplateMessage()));
+                    });
+
+            //javaMailSender.send(mailMessage);
```
[the remark this answers is not in the document]
Comment (Aayushchy) [reply src/main/java/com/esewa/usermanagement/service/EmailServiceImpl.java:53]: Created a separate boolean flag sendEmail in property file

File: src/test/java/com/esewa/usermanagement/service/AdminServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    @Mock
+    private EmailService emailService;
+
+    @Mock
+    private RegistrationLogService registrationLogService;
+    @InjectMocks
+    private AdminServiceImpl adminService;
+
+    @Test
+    void givenValidUser_whenRegisterUser_thenSucceed() {
```
[the remark this answers is not in the document]
Comment (Aayushchy) [reply src/test/java/com/esewa/usermanagement/service/AdminServiceTest.java:34]: Renamed to meaningful name

File: src/test/java/com/esewa/usermanagement/service/AdminServiceTest.java (hunk 1 above)
Comment (Aayushchy) [reply src/test/java/com/esewa/usermanagement/service/AdminServiceTest.java:72]: Done

File: src/test/java/com/esewa/usermanagement/service/UserServiceTest.java (hunk 2 above)
Comment (Aayushchy) [reply src/test/java/com/esewa/usermanagement/service/UserServiceTest.java:82]: Done

**Review by devintern-esewa — APPROVED** [review]
Verified

**Pull request merged**